index.coffee

```diff
@@ -1,60 +1,161 @@
-through = require 'through'
+stream = require 'readable-stream'
+throttle = require 'lodash.throttle'
 
 module.exports = (requestModule, requestOpts, opts) ->
-	res = new Resumable(requestModule, requestOpts, opts)
-	return res.resume()
-
-class Resumable
-	constructor: (requestModule, requestOpts, opts) ->
-		@stream = through (data) =>
-			@bytesRead += data.length
-			@stream.emit('data', data)
+	return new ResumableRequest(requestModule, requestOpts, opts)
+
+module.exports.defaults = (defaults = {}) ->
+	(requestModule, requestOpts, opts) ->
+		opts = Object.assign {}, defaults, opts
+		return new ResumableRequest(requestModule, requestOpts, opts)
+
+cloneResponse = (response) ->
+	# creates an `http.IncomingMessage`-like stream, that can also be piped to.
+	s = new stream.PassThrough()
+	[ 'headers', 'httpVersion', 'method', 'rawHeaders', 'statusCode', 'statusMessage' ].forEach (prop) ->
+		s[prop] = response[prop]
+	return s
+
+wrapLegacyStream = (s) ->
+	new stream.Readable().wrap(s)
+
+class ResumableRequest extends stream.Readable
+	constructor: (@requestModule, @requestOpts, opts = {}) ->
+		super()
+
+		@error = null
+		@request = null
+		@response = null
+
 		@bytesRead = 0
+		@bytesTotal = null
 		@retries = 0
-		@maxRetries = opts?.maxRetries ? 10
-		@contentLength = null
-		@requestOpts = requestOpts
-		@requestModule = requestModule
-		@error = null
+		@maxRetries = opts.maxRetries ? 10
+
+		@_retry = throttle(@_retry.bind(this), opts.retryInterval ? 1000)
+		@_reportProgress = throttle(@_reportProgress.bind(this), opts.progressInterval ? 1000)
+
+		@on 'pipe', =>
+			# request can be written to when uploading a resource to a URL.
+			# We do not support that -- resumable request can only be used
+			# to *download* a resource, i.e. only the response is resumable.
+			# piping will fail with a not-so-obvious error as soon as the stream
+			# gets going because we don't implement Writable, so it's detect this
+			# and emit a more appropriate error.
+			@error = new Error('ResumableRequest is not writable')
+			# have to abort asynchronously to allow client code to register 'error' handlers
+			process.nextTick =>
+				@_retry()
+
+		@_request()
+
+	abort: ->
+		@emit('abort')
+		@_end()
+
+	push: (data, encoding) ->
+		@retries = 0 # we got some data, reset number of retries
+		@bytesRead += data.length if data?
+		@_reportProgress()
+		return super(data, encoding)
+
+	_end: ->
+		return if @ended
+		@ended = true
+		doEnd = =>
+			@request.abort()
+			@_retry.cancel()
+			@_reportProgress()
+			@_reportProgress.flush()
+			@_reportProgress.cancel()
+			@emit('end')
+			if not @error?
+				@emit('complete')
+		if @response?
+			@response.end(doEnd)
+		else
+			doEnd()
+
+	_read: -> # noop -- we're manually pushing buffers into self
```

Comment:
To properly handle back-pressure on the receiving end, this should be implemented – otherwise you could have the resumable reading an entire large download into memory without the destination being able to consume it (classic case would be fast network vs. slow disk), and thus exhausting the hosts resources rather quickly.

You could make it "manually" read from the request like this (crude example, but should illustrate the mechanics, I hope):

```js
_read() {

  // NOTE: For brevity I'll leave out checking for the current request,
  // whether it's ended, and whether to kick off a new one, etc.

  var chunk = null
  var chunksRead = 0

  // Read chunks from the source & push them out
  while( chunk = this.currentRequest.read() ) {
    chunksRead++
    // Avoid pushing out more chunks than the destination can / wants to handle
    if( this.push(chunk) === false ) {
      break
    }
  }

  // If no chunks were read, wait until the source becomes readable again
  if( chunksRead === 0 ) {
    this.currentRequest.once( 'readable', () => {
      this._read()
    })
  }

}
```

Note that the `_read()` method will only be called again by the stream internals once at least one buffer has been pushed out with `this.push(buffer)` from the last call.